spec/consensus/light-client.md

```diff
@@ -316,3 +316,70 @@ func Backwards(h1,h2) bool {
   return (hash(h2) == old.Header.hash)
  }
 ```
+
+### Failure Detection Module
+
+The light client must be able to detect failures in the trust assumptions.
+Ideally, it should never trust a header, which was forged by malicious
+validators.
+
+### Connectivity requirement
+
+To meet the above requirement, it must be connected to at least one honest
+full node. This is sufficient so long as there's no fork on the main chain.
+
+If there is a fork on the main chain, it means that two full nodes have decided
+on two different headers for the same height. Then the requirement would at
+least require that the lite client is **connected to one honest full node on
+each branch of the fork**.
+
+_Assumption_: no more than `2f` of the nodes are malicious (`f+1` are honest).
+otherwise, they could create more forks.
+
+In practice, this means connecting to 1 or more geographically distributed full
+nodes (called witnesses), which belong to different companies. Note this number
+does not include the primary full node (called primary), which is used for
+obtaining new headers.
+
+_Remark_: we can't guarantee all witnesses won't follow the same branch of a
+fork. to guarantee that, we'll need the light client to be connected to +1/3 of
+nodes, which is impractical.
```

Comment:
http://www.nooooooooooooooo.com/

Reply:
Hard guarantees will be super expensive (the lite client has to talk to all full nodes). We looked a bit into probabilistic guarantees, but this may need involved analysis, which we have not done yet. I guess for now the detector should just connect to some witnesses, and replace witnesses it observes problems with them.

Reply:
Expensive how? If latency; requests could be dispatched in parallel and thus the typical latency would be the 33rd percentile round-trip latency, which should be on the order of 100 ms. For something as important as this, isn't that a price worth paying?

Reply:
Related: in order to verify that a sufficient share of the validator set that signed h1 also signed h2, will we not need to gather the public keys of the validators anyway?

Reply:
> in order to verify that a sufficient share of the validator set that signed h1 also signed h2, will we not need to gather the public keys of the validators anyway?

we will need them, but these can be fetched from one node (primary)

Reply:
> Expensive how?

in terms of load. suppose there are 300 vals on Cosmos network. 1/3 = 100. suppose there are 10_000 mobile wallets running a light client. this results in 10_000 requests to 100 nodes every 6 seconds.

Reply:
> in terms of load

Since block headers are immutable I'm thinking they can probably be cached quite heavily on the server-side, so this would basically amount to in-memory hashmap lookups?

Reply:
we don't need headers per se. we only need their hashes to compare them.

Reply:
note 2: current TM full node has an upper bound on peers number - 50

a simple way for us to supply witnesses to light client can be accepting Tendermint's P2P address book (address_book.json) file and using it to fill witnesses.

Reply:
We can't guarantee connectivity to any fraction of nodes or validators. What we're interested in is essentially connectivity to a correct validator, but since they tend to be hidden behind full nodes ("sentries") we can't guarantee that either. So it's not even a matter of performance, the structure of the network just makes it impossible :)